thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,163 @@
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef TRACESLAVE_HPP
+#define TRACESLAVE_HPP
+
+#include "slave.hpp"
+#include "jqueue.tpp"
+#include "jthread.hpp"
+#include "limits.h"
+
+
+class CTracingThorDataLink :  implements CInterfaceOf<IThorDataLink>
+{
+private:
+    class TraceQueue
+    {
+    private:
+        void dump(MemoryBuffer &mb, IHThorArg *helper) const
+        {
+            // NOTE - we don't output every item in the list to be 
+            // consistent with the other dump() implementation
+            // Note - whilst dumping, does not clear ring buffer
+            // so that it is available later
+            unsigned rowPos = rowHeadp;
+            for (unsigned n=1; n<traceRowBufferSize;++n)
+            {
+                if (++rowPos==traceRowBufferSize) rowPos = 0;  // NOTE - skipping first item intentional
+                OwnedConstThorRow row (rowBuffer[rowPos]);
+                if (!row) continue;
+
+                CommonXmlWriter xmlwrite(XWFnoindent|XWFtrim);
+                helper->queryOutputMeta()->toXML((const byte*)(row.get()), xmlwrite);
+                VStringBuffer trace("<Row>%s</Row>", xmlwrite.str());
+                mb.append(trace.str());
+            }
+        }
+    public:
+        OwnedConstThorRow *rowBuffer;
+        unsigned rowHeadp;
+        TraceQueue()
+        {
+            rowBuffer = NULL;
+            rowHeadp = 0;
+        }
+        ~TraceQueue()
+        {
+            delete [] rowBuffer;
+        }
+        void init()
+        {
+            if (!rowBuffer)
+            {
+                rowBuffer = new OwnedConstThorRow[traceRowBufferSize];
+                rowHeadp = 0;
+            }
+        }
+        void enqueue(const void *row)
+        {
+            // NOTE - updates then sets rowHeadp (atomically) so that the code
+            // below (ignoring oldest record) is safe
+            rowBuffer[rowHeadp].set(row);
+            unsigned nextHead = rowHeadp + 1;
+            if (nextHead==traceRowBufferSize)
+                nextHead = 0;
+            rowHeadp = nextHead;
+        }
+        // This dump() makes use of pass-through buffer to ensure if it is called again
+        // then previous rows still available
+        void dump(MemoryBuffer &mb, TraceQueue &rowLogCache, IHThorArg *helper)
+        {
+            // NOTE - we don't output every item in the list as one
+            // COULD have been updated after we swapped the buffers
```

Comment:
>We do not know that the items in the non-active buffer are immediately before the ones in the active buffer - the active buffer may have cycled a number of times so that there is a gap in the sequence between the active and non-active buffers.

That's true, but if that's true, the non-active buffer that is about to be swapped out and added to the tracingQueue will be full, so that the extra row added (that as you point out may be out of date / old, with many since) will be roll off as the new swapped out buffer rows are added to the tracing queue.

It's only if there active queue being swapped out is partially full that it really matters. 

I think this applies what ever the rate.